MOEAD Integration

AiraComet opened the pull request.

UfukAltiparmakoglu committed:
Adding a sample problem TAUQC; working, but not finished yet!

UfukAltiparmakoglu committed:
Update Readme.md

Adding newlines; Minor correction;

kaiamann committed:
Added a "playground" problem

kaiamann committed:
First draft of interfaces

Commit:
Merge branch 'master' of github.com:kaiamann/opt4j

Commit:
Implemented WeigthVector

kaiamann committed:
Merge pull request #1 from loreson/WeightVector

Weight vector and corresponding tests

Commit:
Added Decomposition, SimplexFill and Selection.

Also added the concrete implementations SimplexFillRandom and MaximallySparseSelection.
The DecompositionDefault fills a simplex with N*100 WeightVectors, and selects N  from them via
MaximallySparseSelection. This results in the  MSS-R algorithm.
Possible improvements: Use Das and Dennis filling for MSS-D instead

Commit:
increased number of weightVectors in DecompositionDefaultTest to prevent failing due to bad luck

Commit:
cosmetic changes:
stars in comment, removed unused imports, use <> for constructor calls,
use isEmpty instead of vectror.size==0, removed superflous whitespace

Commit:
Correct use of RNG

use SecureRandom instead of Random, for better quality of Numbers,
make the RNG a class field and reuse it between method calls, to make sonarqube happy
(no actual effect, as in producton SimplexFill.fill most likely will be called only once per object

Commit:
reverted to util.Random due to abysmall performance

Commit:
Refactoring:

Extracted Dimension checking from select to reduce complexity of the select method

Commit:
Merge pull request #1 from loreson/Decomposition

Decomposition

kaiamann committed:
Merge pull request #2 from loreson/Decomposition

Decomposition

kaiamann committed:
Added neighborhood creation via euclidian distance and through random selection

Commit:
first ´draft of the MOEAD. Also typechanges in Repair and PopulationInitialization

Commit:
Added possibility to create multiple Solutions per mating call in MOEAD

UfukAltiparmakoglu committed:
Adding implementation and test for Selector; Minor change in interface;

UfukAltiparmakoglu committed:
Merge branch 'master' of https://github.com/UfukAltiparmakoglu/opt4j

UfukAltiparmakoglu committed:
Overhauling code according to sonarqube; Adding more tests; Adding assertions;

UfukAltiparmakoglu committed:
Further improving code and test;

Commit:
new test cases

Commit:
cleaned up includes

Commit:
Fix comments, add google includes

Commit:
use weightVector.euclidianDistance

kaiamann committed:
Reworked the Neighborhood creation:
- Added an interface to enable switching distance/similarity computation used in creating the neighborhood.
- Added the cosine similarity as such a distance/similarity computation option.
- Removed the random generation of neighborhoods.
Updated test cases for the changed classes accordingly.

UfukAltiparmakoglu committed:
Removing old files and preparing for pull request;

Commit:
Merge branch 'master' into master

Commit:
Merge pull request #2 from UfukAltiparmakoglu/master

Implemented Selector class

Commit:
Added numberOfParents and removed reference Points

Commit:
Merge pull request #3 from kaiamann/master

Neigborhood Creation

UfukAltiparmakoglu committed:
Minor improvements and new test methods added to WeightVector class;

kaiamann committed:
deleted old files

Commit:
Merge pull request #4 from kaiamann/master

deleted old files

UfukAltiparmakoglu committed:
Forgot to merge with master;

UfukAltiparmakoglu committed:
Merge branch 'master' into master

Commit:
Merge pull request #5 from UfukAltiparmakoglu/master

Edited WeightVector and WeightVectorTest

Commit:
Running MOEAD. Added default Implementations and some addition to the glue class

UfukAltiparmakoglu committed:
Fixing bug, where not all parents are returned when neighbourhoodsize and numberofparents are equal;

UfukAltiparmakoglu committed:
Removing duplicate line in build.gradle;

Commit:
ThreeObjectiveTest

Commit:
Revert "use weightVector.euclidianDistance"

This reverts commit c4b94b25b4fae282158ace78d5f0fa62f4f5bd0d.

Commit:
use Mersenne Twister, improved comments, added @ImplementedBy annotations

Commit:
Merge pull request #6 from UfukAltiparmakoglu/master

Minor fixes

Commit:
Merge pull request #7 from loreson/Decomposition

Decomposition

Commit:
Updated Comments and deleted Oilfields

BossSeminarFAU commented on opt4j-core/src/main/java/org/opt4j/core/genotype/BooleanGenotype.java:
```diff
@@ -87,4 +87,15 @@ public void init(Random random, int n) {
 		}
 	}
+
+	public void init(boolean[] a) {
```
Your pull request should not contain any changes to the original Opt4J code. Even in cases where the additions make sense, they should be part of a dedicated pull request and not be done together with the PR contributing the new optimizer. This is point A. I will reference it in other comments.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/EvolutionaryAlgorithm.java:
```diff
@@ -126,6 +126,7 @@ public EvolutionaryAlgorithm(
 	@Override
 	public void initialize() {
 		selector.init(alpha + lambda);
+		System.out.println(population.isEmpty());
```
See point A

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/Decomposition.java:
```diff
@@ -0,0 +1,50 @@
+	 * @param N
+	 *            the number of subproblems (number of weight vectors)
+	 * @param m
+     *            the numbe of objectives (number of entries per weight vector)
```
numbeR

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/MaximallySparseSelection.java:
```diff
@@ -0,0 +1,110 @@
+* of distances to all previously selected vectors.
+* @author Christian Vögl
+*/
+class MaximallySparseSelection implements Selection
```
Very fancy, I like it :)

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/MultiobjectiveEvolutionaryAlgorithm.java:
```diff
@@ -0,0 +1,245 @@
+
+	protected final int n;
+
+	protected final int t;
```
Could you pick a better name here? You should name the variables so that a reader can understand what they do without looking into the paper. Will also make the code more understandable.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/MultiobjectiveEvolutionaryAlgorithm.java:
```diff
@@ -0,0 +1,245 @@
+	public void next() throws TerminationException {
+		completer.complete(population);	
+		for( int i = 0; i < n; i++) {
+			// Step 2.1) Reproduction
```
All the steps in this method need better comments. Again, please do not assume that the reader of the code knows the paper (or even has access to it).

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/MultiobjectiveEvolutionaryAlgorithmModule.java:
```diff
@@ -0,0 +1,263 @@
+	protected int generations = 100;
+
+	@Constant(value = "m", namespace = MultiobjectiveEvolutionaryAlgorithm.class)
+	@Info("The number of objectives per subproblem")
```
This is nice. Nevertheless, giving the stuff descriptive names will make it even better

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/MultiobjectiveEvolutionaryAlgorithmModule.java:
```diff
@@ -0,0 +1,263 @@
+		 * Use a constant crossover rate.
+		 */
+		CONSTANT;
+	}
```
That's an ugly copy paste. Come on, the enum is public. You can use it without copying it here

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/NeighborhoodCreation.java:
```diff
@@ -0,0 +1,110 @@
+ */
+public class NeighborhoodCreation {
+
+    SimilarityMeasure sm;
```
Spontan sieht es aus, als das final sein sollte. Falls du nicht explizit willst, dass man von außen darauf zugreifen will, solltest du es auch private machen.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/NeighborhoodCreation.java:
```diff
@@ -0,0 +1,110 @@
+        int numCandidates = candidates.size();
+
+        // Catch faulty input
+        if (numCandidates < neighborhoodSize) {
+            throw new IllegalArgumentException("Cannot create a neighborhood of size " + neighborhoodSize + "with only "
+                    + numCandidates + "subproblems!");
+        }
+        if (v == null) {
+            throw new IllegalArgumentException("Reference weight vector cannot be null!");
+        }
+        if (neighborhoodSize <= 0) {
+            throw new IllegalArgumentException("Neighborhood size cannot be equal or smaller than 0!");
+        }
+
```
Alles SEHR NICE

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/NeighborhoodCreation.java:
```diff
@@ -0,0 +1,110 @@
+            distances.put(i, distance);
+        }
+
+        // from https://www.javacodegeeks.com/2017/09/java-8-sorting-hashmap-values-ascending-descending-order.html
+        // Sorts the hasmap w.r.t ascending distances
```
sehr guter comment

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/PopulationInitialization.java:
```diff
@@ -0,0 +1,35 @@
+import org.opt4j.core.optimizer.Population;
+
+
+public interface PopulationInitialization {
```
Comment der sagt, was das Ding macht fehlt

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/RandomSelector.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class RandomSelector implements Selector {
```
add Class Comment

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/RandomSelector.java:
```diff
@@ -0,0 +1,56 @@
+			throw new IllegalArgumentException("Provided neighbourhood array is smaller than 2!");
+		}
+	}
+}
```
Es ist gut, dass du die Sachen überprüfst. Nenn die Methoden aber bitte nicht Assert. Das ist eine andere Technik, bei der die Checks nur ausgeführt werden, falls man das Programm mit besonderen Settings ausführt.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/Repair.java:
```diff
@@ -0,0 +1,37 @@
+import com.google.inject.ImplementedBy;
+
+@ImplementedBy(DefaultRepair.class)
+public interface Repair {
```
Class comment

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/Repair.java:
```diff
@@ -0,0 +1,37 @@
+@ImplementedBy(DefaultRepair.class)
+public interface Repair {
+
+    /* Repairs and improves an Individual regarding its objectives
```
Typischerweise wird eine Reparatur verwendet, um falsch gebaute Individuen (die nicht evaluiert werden können) zu reparieren. Da geht es meistens darum, sie überhaupt evaluierbar zu machen, nicht darum, ihre Objectives zu verbessern.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/Selector.java:
```diff
@@ -0,0 +1,40 @@
+import com.google.inject.ImplementedBy;
+
+@ImplementedBy(RandomSelector.class)
+public interface Selector {
```
Class Comment missing

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/Selector.java:
```diff
@@ -0,0 +1,40 @@
+     * @param neighbourhood - the indices of the neighbourhood array
+     * @param numberOfParents - the number of parents
+     * 
+     * @return collection of parents
```
parent indices, not parents

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/SimplexFill.java:
```diff
@@ -0,0 +1,39 @@
+
+
+@ImplementedBy(SimplexFillRandom.class)
+interface SimplexFill
```
Class comment missing

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/SimplexFillRandom.java:
```diff
@@ -0,0 +1,84 @@
+* A. Jaszkiewicz, "On the performance of multiple-objective genetic local
+* search on the 0/1 knapsack problem - a comparative experiment," in IEEE
+* Transactions on Evolutionary Computation, vol. 6, no. 4, pp. 402-412, Aug.
+* 2002.
```
Sehr nice

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/SimplexFillRandom.java:
```diff
@@ -0,0 +1,84 @@
+            }
+            points.add(new WeightVector(values));
+        }
+        return points;
```
kannst du hier Kommentare hinzufügen, die zumindest ungefähr beschreiben, was da passiert? Du solltest generell nicht davon ausgehen, dass jeder, der mit dem Code arbeitet, Zugriff auf das Paper hat, um es dort nachzuschauen.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/WeightVector.java:
```diff
@@ -0,0 +1,78 @@
+ * SOFTWARE.
+ *******************************************************************************/
+
+package org.opt4j.optimizers.ea.moead;
```
Add class and method comments.

BossSeminarFAU commented on opt4j-optimizers/src/main/java/org/opt4j/optimizers/ea/moead/package-info.java:
```diff
@@ -0,0 +1 @@
+package org.opt4j.optimizers.ea.moead;
```
Add a comment describing the contents of the package.

BossSeminarFAU commented on opt4j-optimizers/src/test/java/org/opt4j/optimizers/ea/moead/WeightVectorTest.java:
```diff
@@ -0,0 +1,108 @@
+import java.lang.Math;
+
+public class WeightVectorTest {
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
```
Fancy pancy :) . Kannte ich nicht

UfukAltiparmakoglu committed:
Reverting build.gradle changes; Adding class comments to Selector and RandomSelector class; Renaming assertion methods;

Commit:
More descriptive variables and comments

Commit:
merge

Commit:
Improved variable names and comments

kaiamann committed:
Added comments to WeightVector
Adjusted weight vector access in EuclideanDistance and CosineSimilarity
Removed unnecessary constructor in BooleanGenotype

kaiamann committed:
Merge branch 'master' of https://github.com/AiraComet/opt4j

Commit:
better variable names

Commit:
better variable names

Commit:
renamend MOEAD and updated default parameters

Commit:
added changeable overfill

Commit:
trailing whitespace

kaiamann committed:
Added possibility to select CosineSimilarity for neighborhood creation in GUI

Commit:
made overfill a parameter

kaiamann committed:
Merge branch 'master' of https://github.com/AiraComet/opt4j

kaiamann committed:
fixed MOEAD

kaiamann committed:
Added reference

kaiamann committed:
Adjusted tests and removed code smells

kaiamann committed:
Removed unused imports

Commit:
added Mockito test for MOEAD

Commit:
Merged Overfill and similarity measure, Added overfill test and updated comments

ghost commented:
New and improved Pull Request. 
Realized your remarks as good as we could :D

BossSeminarFAU reviewed (commented):
Looks okay

BossSeminarFAU reviewed (commented):
ok
